src/main/java/frc/robot/subsystems/AbstractCamera.java

```diff
@@ -0,0 +1,88 @@
+import org.photonvision.PhotonUtils;
+import org.photonvision.targeting.PhotonTrackedTarget;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.trajectory.Trajectory;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+// Base class of camera subsystem
+public abstract class AbstractCamera extends SubsystemBase {
+    protected PhotonCamera camera;
+    protected int pipelineIndex;
+    protected double cameraHeight, cameraPitch, targetHeight;
+    protected Translation2d cameraOffset;
+
+    /**
+     * Constructs the camera
+     *
+     * @param cameraName The nickname of the camera.
+     * @param pipelineIndex The index of the pipeline
+     * @param cameraHeight The height of the camera above the ground, in meters
+     * @param cameraPitch The pitch of the camera above the ground, in degrees
+     * @param cameraOffset The offset of the camera from the center of the robot as a translation, in meters
+     * @param targetHeight The height of the target, in meters
+     */    
+    public AbstractCamera(String cameraName, int pipelineIndex, double cameraHeight, double cameraPitch, Translation2d cameraOffset, double targetHeight) {
+        this.camera = new PhotonCamera(cameraName);
+        this.camera.setPipelineIndex(pipelineIndex);
+
+        this.cameraHeight = cameraHeight;
+        this.cameraPitch = cameraPitch;
+        this.cameraOffset = cameraOffset;
+        this.targetHeight = targetHeight;
+    }
+
+    /**
+     * Returns whether the camera has targets.
+     *
+     * @return If the camera has targets.
+     */
+    public boolean hasTargets() {
+        return this.camera.getLatestResult().hasTargets();
+    }
+
+    /**
+     * Finds the camera's best target
+     */
+    public PhotonTrackedTarget getBestTarget() {
+        var result = this.camera.getLatestResult();
+        if (!result.hasTargets()) {
+            return null;
+        }
+        return result.getBestTarget();
+    }
+
+    /**
+     * Gets the distance between the camera and the target
+     * 
+     * @param target The target to get the distance from
+     * @return The distance between the camera and the target, in meters
+    */
+    public double getDistance(PhotonTrackedTarget target) {
+        return PhotonUtils.calculateDistanceToTargetMeters(this.cameraHeight, this.targetHeight, this.cameraPitch, target.getPitch());
+    }
+
+    /**
+     * Finds the translation to the camera's current target
+     * 
+     * @return The translation to the camera's best target
+     */
+    public Translation2d getRelativeTranslation() {
+        var target = this.getBestTarget();
+        double distance = this.getDistance(target);
+
+        Translation2d translation = PhotonUtils.estimateCameraToTargetTranslation(
+            distance, Rotation2d.fromDegrees(-target.getYaw())
+        );
+        translation.plus(this.cameraOffset);
+        return translation;
+    }
```

Comment:
Right now, if you call `getRelativeTranslation` and there are no targets detected, there is going to be a null pointer exception. To prevent that from happening when we call this method in a command, we need to call the `hasTargets` method first to verify that there are targets. The only problem is that they both call the `getLatestResult` method individually, meaning that we could call `hasTargets` verify that that there are targets and then with `getRelativeTranslation` get no targets and end up with a null pointer exception. To remedy this problem, I would recommend that you store the return value of `getLatestResult` as an instance variable and then update it in the overridden `periodic` method of the subsystem, which is called at the beginning of every cycle of the program. You can then use the stored value instead of calling the `getLatestResult` method to ensure that you are using the same result throughout the execution of the program and avoid unwanted errors.

Reply:
I think this code is fine because this class' `getDistance` method calls `hasTargets` on its own copy of the latest result as opposed to using the instance's `hasTargets` method. As long as we add a null check it should work. Unless I'm missing something?

Reply:
Yes, we could add a null check, but in the future when this class is extended `getTrajectory` is going to call `getRelativeTranslation` and will have to run a null check there as well. I think that it would be more efficient to call `hasTargets` from the command that uses these methods to verify that there are targets beforehand. Either way works though.

Reply:
If we get a result periodically, there would still be cases where there's no targets. We can't just store non-null values, since there would be calculations are done with outdated data

Reply:
I am talking about storing the pipeline result as an instance variable at the beginning of each cycle of the program regardless of whether or not there are targets. And then within the command that is calling all of these methods, only call the `getTrajectory` method and run the command once it has been verified that there are targets in the result stored in the instance variable.

A couple of null checks would work as well though.

Reply:
This is a great place to start writing some unit tests. We can set up scenarios for each of these cases in there and verify it acts appropriately.

I don't have any objections to merging this as it is but I think we should come back and write a comprehensive unit test for this class in another PR.

Reply:
Alright, that sounds good.